### PR TITLE
Optimize better with maximum array length

### PR DESCRIPTION
**Pull request by BruceForstall**
The maximum array length is 0x7FFFFFC7 (see
https://github.com/dotnet/runtime/pull/43301), which is slightly less than
the largest 32-bit signed integer. Use this value in range check and assertion
prop to optimize better. This is due do knowing that a value smaller than
the maximum array length plus a small constant will not overflow. This leads
to being able to remove some bounds checks.

### CODE REVIEW

**Comment by BruceForstall** [comment]
@AndyAyersMS PTAL
cc @SingleAccretion @dotnet/jit-contrib

File: src/coreclr/jit/assertionprop.cpp [hunk 1]
```diff
@@ -145,7 +158,7 @@ bool IntegralRange::Contains(int64_t value) const
             return {SymbolicIntegerValue::Zero, SymbolicIntegerValue::One};
 
         case GT_ARR_LENGTH:
-            return {SymbolicIntegerValue::Zero, SymbolicIntegerValue::IntMax};
+            return {SymbolicIntegerValue::Zero, SymbolicIntegerValue::ArrayLenMax};
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/coreclr/jit/assertionprop.cpp:161]: Yeah, GC fails fast for such objects (e.g. `new string('\0', int.MaxValue);`)

File: src/coreclr/jit/jit.h [hunk 2]
```diff
@@ -851,6 +851,14 @@ extern "C" UINT32 WINAPI getLikelyClasses(LikelyClassRecord*
                                           BYTE*                                  pInstrumentationData,
                                           int32_t                                ilOffset);
 
+/*****************************************************************************/
+
+// The maximum array length is defined to be slightly less than the maximum signed 32-bit integer.
+// See https://github.com/dotnet/runtime/pull/43301, https://msdn.microsoft.com/en-us/windows/apps/hh285054.aspx.
+// CLR throws IDS_EE_ARRAY_DIMENSIONS_EXCEEDED if array length is too large. The practical limit is likely smaller
+// due to memory fragmentation.
+#define ARRLEN_MAX (0x7FFFFFC7)
```
[the remark this answers is not in the document]
Comment (BruceForstall) [reply src/coreclr/jit/jit.h]: Thanks for pointing that out. Fixed.

[the remark this answers is not in the document]
Comment (BruceForstall) [reply src/coreclr/jit/jit.h]: @SingleAccretion Does this observation imply an issue with this change, or you're just pointing out to be careful with range bounds?

[the remark this answers is not in the document]
Comment (BruceForstall) [reply src/coreclr/jit/jit.h]: Maybe I'm missing something. When I compile your test (specifically:
```
using System;
using System.Runtime.CompilerServices;
namespace testns
{
    class testclass
    {
    [MethodImpl(MethodImplOptions.NoInlining)]
    private static int Problem(Span<int> a)
    {
        int sum = 0;
        for (int i = 0; i < a.Length; i += 2) // 'i' will overflow for "a.Length == int.MaxValue".
        {
            sum += a[i];
        }
        return sum;
    }
    static void Main(string[] args)
    {
        int[] a = new int[10] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
        int s = Problem(a);
        Console.WriteLine(s);
    }
    }
}
```
), I see `a.Length` expanded as
```
CALL int System.Span`1[Int32][System.Int32].get_Length
```
whereas this PR only affects `GT_ARR_LENGTH` nodes, and there are none in the test, so I don't see any diff for this case.

[the remark this answers is not in the document]
Comment (SingleAccretion) [reply src/coreclr/jit/jit.h]: > so I don't see any diff for this case

That is very surprising to me, there should be a diff with the removed check.

As a reference, I have checked out the change locally and obtained two dumps for `Problem`, one for the baseline and one for the diff: [baselog.txt](https://github.com/dotnet/runtime/files/8779973/baselog.txt), [difflog.txt](https://github.com/dotnet/runtime/files/8779982/difflog.txt), and I do see we remove the check in the diff, which corresponds to this change in range check's logic:
```diff
            int tmp = GetArrLength(limit.vn);
            if (tmp <= 0)
            {
-               tmp = ARRLEN_MAX;
+               tmp = CORINFO_Array_MaxLength;
            }
```

[the remark this answers is not in the document]
Comment (BruceForstall) [reply src/coreclr/jit/jit.h]: Ok, it was operator error: using a Debug S.P.C.dll prevents cross-module inlining, and I normally use a Debug build (if at all possible). Running with Checked (and a Debug JIT) repros the problem.

Seems like VN/AssertionProp doesn't keep array and non-array GT_BOUNDS_CHECK usage distinct. Maybe I can revert this line for now and come back to it later.

**Comment by BruceForstall** [comment]
@SingleAccretion Take a look at the latest update: it's only using the array max length if we've got a GT_ARR_LENGTH. This preserves most of the wins, but your test case is zero diff.

**Review by SingleAccretion — APPROVED** [review]
LGTM.

**Review by EgorBo — APPROVED** [review]

**Comment by BruceForstall** [comment]
@SingleAccretion Thanks for the great feedback!

**Comment by EgorBo** [comment]
Improvements on x64: https://github.com/dotnet/perf-autofiling-issues/issues/5618

**Comment by stephentoub** [comment]
> Improvements on x64: https://github.com/dotnet/perf-autofiling-issues/issues/5618

I suspect most of those improvements are actually due to https://github.com/dotnet/runtime/pull/69839.

**Comment by EgorBo** [comment]
> > Improvements on x64: [dotnet/perf-autofiling-issues#5618](https://github.com/dotnet/perf-autofiling-issues/issues/5618)
> 
> I suspect most of those improvements are actually due to #69839.

Oops 🙂

**Pull request merged**